### PR TITLE
Do not allow private members to be referred to in cref across assemblies

### LINKED ISSUE #826: cref should not be able to access private fields in other assemblies
This is a continued discussion from issue #77 

Today the binding which occurs during a `cref` lookup does not respect accessibility.  This is a back compat behavior to the native compiler which makes sense to a degree.  The ability to reference private fields which are defined in the same assembly does have documentation value.  

But I think the dangers of allowing references to private fields across assembly boundaries outweighs the documentation benefit because it creates the following issues:

- It should force a change in how tools generate reference assemblies.  Today they typically exclude private data.  This prevents them from being perfect stand ins for non-reference assemblies.  For example today I cannot add a `cref` to `string.m_firstChar` because the ref assembly excludes it.  
- It opens up the possibility that private member name changes can cause compilation errors in unrelated assemblies.  Private member rename should not be a back compat break.  
- It reduces the set of refactorings that can be safely done intra-project which slows down our overall refactoring story.
- It opens up the possibility that adding a new private member creates a compilation error (could create an ambiguity with other accessible members used by a `cref`).  

Limiting the accessibility suppression to be intra-assembly only would mitigate these concerns.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #826 

This is a preliminary PR. I will add both positive and negative tests before submitting.

### CODE REVIEW

**Comment by gafter** [comment]
@japredpar @agocke @amcasey Please have a look for implementation strategy. Tests to be added.

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs [hunk 1]
```diff
@@ -1097,6 +1097,12 @@ internal SingleLookupResult CheckViability(Symbol symbol, int arity, LookupOptio
             {
                 return LookupResult.NotInvocable(unwrappedSymbol, symbol, diagnose);
             }
+            else if (InCref && !this.IsCrefAccessible(unwrappedSymbol))
+            {
+                var unwrappedSymbols = ImmutableArray.Create<Symbol>(unwrappedSymbol);
```
Comment (amcasey) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs]: I suppose it doesn't really matter, but we don't appear to use this if diagnose is false.  Maybe just inline the expression?

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs [hunk 2]
```diff
@@ -1246,6 +1253,28 @@ private static TypeSymbol RefineAccessThroughType(LookupOptions options, TypeSym
...
+        private bool IsCrefAccessible(Symbol symbol)
+        {
+            return !IsEffectivelyPrivate(symbol) || symbol.ContainingAssembly == this.Compilation.Assembly;
+        }
+
+        private static bool IsEffectivelyPrivate(Symbol symbol)
+        {
+            for (Symbol s = symbol; (object)s != null; s = s.ContainingSymbol)
+            {
+                if (s.DeclaredAccessibility == Accessibility.Private)
```
Comment (amcasey) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs]: Aren't there some symbols that don't have DeclaredAccessibility?  Are any of them effectively private?

**Comment by amcasey** [comment]
:+1:

**Comment by amcasey** [comment]
There should probably be a comment capturing the rationale somewhere.

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs (hunk 2 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs]: No, things like namespaces report their accessibility as public. I don't think there are things that are effectively private that don't report such.

Comment (amcasey) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs]: What do explicit interface implementations report?

Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs]: I don't think there is a way to refer to them in cref.

**Comment by jaredpar** [comment]
:+1:

File: src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs (hunk 2 above)
Comment (sharwell) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs]: You can, but it's not pretty. The following reference should link to [`Dictionary<TKey, TValue>.ICollection<KeyValuePair<TKey, TValue>>.Contains`](https://msdn.microsoft.com/en-us/library/cc672218.aspx).

```xml
<see cref="M:System.Collections.Generic.Dictionary`2.System#Collections#Generic#ICollection{System.Collections.Generic.KeyValuePair{`0,`1}}#Contains(System.Collections.Generic.KeyValuePair{`0,`1})"/>
```

However, due to an issue in MTPS, this reference won't actually get resolved. MTPS expects the reference to have the following form:

```xml
<see cref="M:System.Collections.Generic.Dictionary`2.System#Collections#Generic#ICollection{T}#Contains(System.Collections.Generic.KeyValuePair{`0,`1})"/>
```

This means you could actually have two methods with identical signatures if a type explicitly implemented two different parameterizations of the same generic type, and that type declared a method which did not have any arguments. [Despite my best efforts](http://stackoverflow.com/q/22416181/138304) I was not able to locate a situation where this actually led to a presentation bug.

**Pull request merged**